readthedocs/proxito/views.py

```diff
@@ -0,0 +1,247 @@
+    """
+
+    @wraps(view_func)
+    def inner_view(  # noqa
+            request, subproject=None, subproject_slug=None, *args, **kwargs
+    ):
+        if subproject is None and subproject_slug:
+            # Try to fetch by subproject alias first, otherwise we might end up
+            # redirected to an unrelated project.
+            try:
+                # Depends on a project passed into kwargs
+                rel = ProjectRelationship.objects.get(
+                    parent=kwargs['project'],
+                    alias=subproject_slug,
+                )
+                subproject = rel.child
+            except (ProjectRelationship.DoesNotExist, KeyError):
+                subproject = get_object_or_404(Project, slug=subproject_slug)
+        return view_func(request, subproject=subproject, *args, **kwargs)
+
+    return inner_view
+
+
+def map_project_slug(view_func):
+    """
+    A decorator that maps a ``project_slug`` URL param into a Project.
+
+    :raises: Http404 if the Project doesn't exist
+
+    .. warning:: Does not take into account any kind of privacy settings.
+    """
+
+    @wraps(view_func)
+    def inner_view(  # noqa
+            request, project=None, project_slug=None, *args, **kwargs
+    ):
+        if project is None:
+            # Get a slug from the request if it can't be found in the URL
+            if not project_slug:
+                project_slug = request.host_project_slug
+                log.debug(
+                    'Inserting project slug from request slug=[%s]',
+                    project_slug
+                )
+            try:
+                project = Project.objects.get(slug=project_slug)
+            except Project.DoesNotExist:
+                raise Http404('Project does not exist.')
+        return view_func(request, project=project, *args, **kwargs)
+
+    return inner_view
+
+
+@map_project_slug
+@map_subproject_slug
+def redirect_page_with_filename(request, project, subproject, filename):  # pylint: disable=unused-argument  # noqa
+    """Redirect /page/file.html to /<default-lang>/<default-version>/file.html."""
+
+    urlparse_result = urlparse(request.get_full_path())
+    return HttpResponseRedirect(
+        resolve(
+            subproject or project,
+            filename=filename,
+            query_params=urlparse_result.query,
+        )
+    )
+
+
+@map_project_slug
+@map_subproject_slug
+def redirect_project_slug(request, project, subproject):  # pylint: disable=unused-argument
+    """Handle / -> /en/latest/ directs on subdomains."""
+    urlparse_result = urlparse(request.get_full_path())
+
+    return HttpResponseRedirect(
+        resolve(
+            subproject or project,
+            query_params=urlparse_result.query,
+        ),
+    )
+
+
+@map_project_slug
+@map_subproject_slug
+def serve_docs(
+        request,
+        project,
+        subproject,
+        lang_slug=None,
+        version_slug=None,
+        filename='',
```

Comment:
Because of the way Django processes the URL regex, it always matches to ''. I could leave it without a default actually, but this seems more explicit?

Reply:
I see --no strong opinion, then.